src/qibosoq/programs/base.py

```diff
@@ -328,24 +345,26 @@ def add_muxed_readout_to_register(self, ro_pulses: List[Pulse]):
         self.set_pulse_registers(ch=gen_ch, style="const", length=length, mask=mask)
 
     def create_mux_ro_dict(self) -> dict:
-        """Create a dictionary containing grouped readout pulses.
+        """Create a dictionary containing readout pulses grouped by start time.
 
         Example of dictionary:
-        { 'start_time_0': [pulse1, pulse2],
-        'start_time_1': [pulse3]}
+        { 0: [pulse1, pulse2], 1: [pulse3]}
         """
         mux_dict = {}
+        len_last_readout = 0
         for pulse in (pulse for pulse in self.sequence if pulse.type == "readout"):
-            if round(pulse.start, 5) not in mux_dict:
+            if pulse.start_delay <= len_last_readout:
+                # add the pulse to the last multiplexed readout
                 if len(mux_dict) > 0:
-                    if (pulse.start - list(mux_dict)[-1]) < 2:  # TODO not 2, but pulses len
-                        mux_dict[round(pulse.start, 5)] = mux_dict[list(mux_dict)[-1]]
-                        del mux_dict[list(mux_dict)[-2]]
-                    else:
-                        mux_dict[round(pulse.start, 5)] = []
+                    mux_dict[list(mux_dict)[-1]].append(pulse)
                 else:
-                    mux_dict[round(pulse.start, 5)] = []
-            mux_dict[round(pulse.start, 5)].append(pulse)
+                    mux_dict[0] = [pulse]
+
+            elif pulse.start_delay > len_last_readout:
+                # add a new multiplexed readout
+                mux_dict[len(mux_dict)] = [pulse]
```

Comment:
Why not a list? You're making many operations that are definitely list-like. This would be just:
```suggestion
                mux.append([pulse])
```

Reply:
If you need to, you can always convert the list to a dictionary before returning.